cannot deserialize from Object value (no delegate- or property-based Creator)

### Describe the bug

This only happens in the native version..

Stacktrace:

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.crunchydata.model.Clusters` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (ByteArrayInputStream); line: 1, column: 2]
	at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1764)
	at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:400)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1209)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1415)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:362)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:195)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2033)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1426)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.JacksonBasicMessageBodyReader.doReadFrom(JacksonBasicMessageBodyReader.java:45)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.JacksonBasicMessageBodyReader.readFrom(JacksonBasicMessageBodyReader.java:34)
	at org.jboss.resteasy.reactive.client.impl.ClientReaderInterceptorContextImpl.proceed(ClientReaderInterceptorContextImpl.java:54)
```


Here is the definition of Clusters

```java
package com.crunchydata.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public class Clusters {
    @JsonProperty("clusters")
    List<Cluster> clusterList;

    public Clusters() {}

    public Clusters(List<Cluster> clusterList) {
        this.clusterList = clusterList;
    }

    public List<Cluster> getClusterList() {
        return clusterList;
    }

    public void setClusterList(List<Cluster> clusterList) {
        this.clusterList = clusterList;
    }
}
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Darwin Daves-MacBook-Pro.local 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64

### Output of `java -version`

OpenJDK 64-Bit Server VM GraalVM CE 21.2.0 (build 16.0.2+7-jvmci-21.2-b08, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.2.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.0

### Additional information

This only happens with the native-image

--- a/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
+++ b/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
@@ -23,6 +23,7 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletionStage;
+import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.regex.Pattern;
 
@@ -84,6 +85,7 @@ import org.jboss.resteasy.reactive.common.model.RestClientInterface;
 import org.jboss.resteasy.reactive.common.processor.AdditionalReaderWriter;
 import org.jboss.resteasy.reactive.common.processor.AdditionalReaders;
 import org.jboss.resteasy.reactive.common.processor.AdditionalWriters;
+import org.jboss.resteasy.reactive.common.processor.EndpointIndexer;
 import org.jboss.resteasy.reactive.common.processor.HashUtil;
 import org.jboss.resteasy.reactive.common.processor.ResteasyReactiveDotNames;
 import org.jboss.resteasy.reactive.common.processor.scanning.ResourceScanningResult;
@@ -106,6 +108,7 @@ import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.RuntimeInitializedClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ServiceProviderBuildItem;
 import io.quarkus.deployment.metrics.MetricsCapabilityBuildItem;
@@ -125,6 +128,7 @@ import io.quarkus.jaxrs.client.reactive.runtime.JaxrsClientReactiveRecorder;
 import io.quarkus.jaxrs.client.reactive.runtime.ToObjectArray;
 import io.quarkus.resteasy.reactive.common.deployment.ApplicationResultBuildItem;
 import io.quarkus.resteasy.reactive.common.deployment.QuarkusFactoryCreator;
+import io.quarkus.resteasy.reactive.common.deployment.QuarkusResteasyReactiveDotNames;
 import io.quarkus.resteasy.reactive.common.deployment.ResourceScanningResultBuildItem;
 import io.quarkus.resteasy.reactive.common.deployment.SerializersUtil;
 import io.quarkus.resteasy.reactive.common.runtime.ResteasyReactiveConfig;
@@ -188,6 +192,7 @@ public class JaxrsClientReactiveProcessor {
             BeanContainerBuildItem beanContainerBuildItem,
             ApplicationResultBuildItem applicationResultBuildItem,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClassBuildItemBuildProducer,
+            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchyBuildItemBuildProducer,
             List<MessageBodyReaderBuildItem> messageBodyReaderBuildItems,
             List<MessageBodyWriterBuildItem> messageBodyWriterBuildItems,
             List<MessageBodyReaderOverrideBuildItem> messageBodyReaderOverrideBuildItems,
@@ -238,6 +243,24 @@ public class JaxrsClientReactiveProcessor {
                 .setHasRuntimeConverters(false)
                 .setDefaultProduces(defaultProducesType)
                 .setSmartDefaultProduces(disableSmartDefaultProduces.isEmpty())
+                .setResourceMethodCallback(new Consumer<EndpointIndexer.ResourceMethodCallbackData>() {
+                    @Override
+                    public void accept(EndpointIndexer.ResourceMethodCallbackData entry) {
+                        MethodInfo method = entry.getMethodInfo();
+                        String source = JaxrsClientReactiveProcessor.class.getSimpleName() + " > " + method.declaringClass()
+                                + "[" + method + "]";
+
+                        reflectiveHierarchyBuildItemBuildProducer.produce(new ReflectiveHierarchyBuildItem.Builder()
+                                .type(method.returnType())
+                                .index(index)
+                                .ignoreTypePredicate(QuarkusResteasyReactiveDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
+                                .ignoreFieldPredicate(QuarkusResteasyReactiveDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
+                                .ignoreMethodPredicate(
+                                        QuarkusResteasyReactiveDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
+                                .source(source)
+                                .build());
+                    }
+                })
                 .build();
 
         boolean observabilityIntegrationNeeded = (capabilities.isPresent(Capability.OPENTELEMETRY_TRACER) ||

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/QuarkusResteasyReactiveDotNames.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/QuarkusResteasyReactiveDotNames.java
@@ -1,7 +1,12 @@
 package io.quarkus.resteasy.reactive.common.deployment;
 
-import org.jboss.jandex.DotName;
+import java.util.function.Predicate;
 
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.FieldInfo;
+import org.jboss.jandex.MethodInfo;
+
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.vertx.core.http.HttpServerRequest;
 import io.vertx.core.http.HttpServerResponse;
 
@@ -9,5 +14,58 @@ public class QuarkusResteasyReactiveDotNames {
 
     public static final DotName HTTP_SERVER_REQUEST = DotName.createSimple(HttpServerRequest.class.getName());
     public static final DotName HTTP_SERVER_RESPONSE = DotName.createSimple(HttpServerResponse.class.getName());
+    public static final DotName JSON_IGNORE = DotName.createSimple("com.fasterxml.jackson.annotation.JsonIgnore");
+    public static final DotName JSONB_TRANSIENT = DotName.createSimple("javax.json.bind.annotation.JsonbTransient");
+    public static final IgnoreTypeForReflectionPredicate IGNORE_TYPE_FOR_REFLECTION_PREDICATE = new IgnoreTypeForReflectionPredicate();
+    public static final IgnoreFieldForReflectionPredicate IGNORE_FIELD_FOR_REFLECTION_PREDICATE = new IgnoreFieldForReflectionPredicate();
+    public static final IgnoreMethodForReflectionPredicate IGNORE_METHOD_FOR_REFLECTION_PREDICATE = new IgnoreMethodForReflectionPredicate();
+    private static final String[] PACKAGES_IGNORED_FOR_REFLECTION = {
+            // JSON-P
+            "javax.json.",
+            // Jackson
+            "com.fasterxml.jackson.databind.",
+            // JAX-RS
+            "javax.ws.rs.",
+            // RESTEasy
+            "org.jboss.resteasy.reactive",
+            // Vert.x JSON layer
+            "io.vertx.core.json.",
+            // Mutiny
+            "io.smallrye.mutiny."
+    };
 
+    private static class IgnoreTypeForReflectionPredicate implements Predicate<DotName> {
+
+        @Override
+        public boolean test(DotName dotName) {
+            if (ReflectiveHierarchyBuildItem.DefaultIgnoreTypePredicate.INSTANCE.test(dotName)) {
+                return true;
+            }
+            String name = dotName.toString();
+            for (String packageName : PACKAGES_IGNORED_FOR_REFLECTION) {
+                if (name.startsWith(packageName)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    private static class IgnoreFieldForReflectionPredicate implements Predicate<FieldInfo> {
+
+        @Override
+        public boolean test(FieldInfo fieldInfo) {
+            return fieldInfo.hasAnnotation(JSON_IGNORE)
+                    || fieldInfo.hasAnnotation(JSONB_TRANSIENT);
+        }
+    }
+
+    private static class IgnoreMethodForReflectionPredicate implements Predicate<MethodInfo> {
+
+        @Override
+        public boolean test(MethodInfo methodInfo) {
+            return methodInfo.hasAnnotation(JSON_IGNORE)
+                    || methodInfo.hasAnnotation(JSONB_TRANSIENT);
+        }
+    }
 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
@@ -96,6 +96,7 @@ import io.quarkus.netty.deployment.MinNettyAllocatorMaxOrderBuildItem;
 import io.quarkus.resteasy.reactive.common.deployment.ApplicationResultBuildItem;
 import io.quarkus.resteasy.reactive.common.deployment.FactoryUtils;
 import io.quarkus.resteasy.reactive.common.deployment.QuarkusFactoryCreator;
+import io.quarkus.resteasy.reactive.common.deployment.QuarkusResteasyReactiveDotNames;
 import io.quarkus.resteasy.reactive.common.deployment.ResourceInterceptorsBuildItem;
 import io.quarkus.resteasy.reactive.common.deployment.ResourceScanningResultBuildItem;
 import io.quarkus.resteasy.reactive.common.deployment.SerializersUtil;
@@ -373,10 +374,10 @@ public class ResteasyReactiveProcessor {
                             reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem.Builder()
                                     .type(method.returnType())
                                     .index(index)
-                                    .ignoreTypePredicate(ResteasyReactiveServerDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
-                                    .ignoreFieldPredicate(ResteasyReactiveServerDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
+                                    .ignoreTypePredicate(QuarkusResteasyReactiveDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
+                                    .ignoreFieldPredicate(QuarkusResteasyReactiveDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
                                     .ignoreMethodPredicate(
-                                            ResteasyReactiveServerDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
+                                            QuarkusResteasyReactiveDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
                                     .source(source)
                                     .build());
 
@@ -387,11 +388,11 @@ public class ResteasyReactiveProcessor {
                                             .type(parameterType)
                                             .index(index)
                                             .ignoreTypePredicate(
-                                                    ResteasyReactiveServerDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
+                                                    QuarkusResteasyReactiveDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
                                             .ignoreFieldPredicate(
-                                                    ResteasyReactiveServerDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
+                                                    QuarkusResteasyReactiveDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
                                             .ignoreMethodPredicate(
-                                                    ResteasyReactiveServerDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
+                                                    QuarkusResteasyReactiveDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
                                             .source(source)
                                             .build());
                                 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveServerDotNames.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveServerDotNames.java
@@ -1,18 +1,13 @@
 package io.quarkus.resteasy.reactive.server.deployment;
 
-import java.util.function.Predicate;
-
 import javax.ws.rs.core.Context;
 
 import org.jboss.jandex.DotName;
-import org.jboss.jandex.FieldInfo;
-import org.jboss.jandex.MethodInfo;
 import org.jboss.resteasy.reactive.server.ServerRequestFilter;
 import org.jboss.resteasy.reactive.server.ServerResponseFilter;
 import org.jboss.resteasy.reactive.server.SimpleResourceInfo;
 import org.jboss.resteasy.reactive.server.spi.ResteasyReactiveContainerRequestContext;
 
-import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.vertx.ext.web.RoutingContext;
 
 public class ResteasyReactiveServerDotNames {
@@ -26,60 +21,4 @@ public class ResteasyReactiveServerDotNames {
     public static final DotName SIMPLIFIED_RESOURCE_INFO = DotName.createSimple(SimpleResourceInfo.class.getName());
     public static final DotName ROUTING_CONTEXT = DotName.createSimple(RoutingContext.class.getName());
 
-    public static final DotName JSON_IGNORE = DotName.createSimple("com.fasterxml.jackson.annotation.JsonIgnore");
-    public static final DotName JSONB_TRANSIENT = DotName.createSimple("javax.json.bind.annotation.JsonbTransient");
-
-    public static final IgnoreTypeForReflectionPredicate IGNORE_TYPE_FOR_REFLECTION_PREDICATE = new IgnoreTypeForReflectionPredicate();
-    public static final IgnoreFieldForReflectionPredicate IGNORE_FIELD_FOR_REFLECTION_PREDICATE = new IgnoreFieldForReflectionPredicate();
-    public static final IgnoreMethodForReflectionPredicate IGNORE_METHOD_FOR_REFLECTION_PREDICATE = new IgnoreMethodForReflectionPredicate();
-
-    private static class IgnoreTypeForReflectionPredicate implements Predicate<DotName> {
-
-        @Override
-        public boolean test(DotName dotName) {
-            if (ReflectiveHierarchyBuildItem.DefaultIgnoreTypePredicate.INSTANCE.test(dotName)) {
-                return true;
-            }
-            String name = dotName.toString();
-            for (String packageName : PACKAGES_IGNORED_FOR_REFLECTION) {
-                if (name.startsWith(packageName)) {
-                    return true;
-                }
-            }
-            return false;
-        }
-    }
-
-    private static class IgnoreFieldForReflectionPredicate implements Predicate<FieldInfo> {
-
-        @Override
-        public boolean test(FieldInfo fieldInfo) {
-            return fieldInfo.hasAnnotation(JSON_IGNORE)
-                    || fieldInfo.hasAnnotation(JSONB_TRANSIENT);
-        }
-    }
-
-    private static class IgnoreMethodForReflectionPredicate implements Predicate<MethodInfo> {
-
-        @Override
-        public boolean test(MethodInfo methodInfo) {
-            return methodInfo.hasAnnotation(JSON_IGNORE)
-                    || methodInfo.hasAnnotation(JSONB_TRANSIENT);
-        }
-    }
-
-    private static final String[] PACKAGES_IGNORED_FOR_REFLECTION = {
-            // JSON-P
-            "javax.json.",
-            // Jackson
-            "com.fasterxml.jackson.databind.",
-            // JAX-RS
-            "javax.ws.rs.",
-            // RESTEasy
-            "org.jboss.resteasy.reactive",
-            // Vert.x JSON layer
-            "io.vertx.core.json.",
-            // Mutiny
-            "io.smallrye.mutiny."
-    };
 }

--- a/integration-tests/rest-client-reactive/src/main/java/io/quarkus/it/rest/client/main/Apple.java
+++ b/integration-tests/rest-client-reactive/src/main/java/io/quarkus/it/rest/client/main/Apple.java
@@ -1,9 +1,5 @@
 package io.quarkus.it.rest.client.main;
 
-import io.quarkus.runtime.annotations.RegisterForReflection;
-
-// TODO get rid of need for this
-@RegisterForReflection
 public class Apple {
     private String cultivar;
 

